src/main/java/org/apache/sysds/runtime/instructions/fed/ParameterizedBuiltinFEDInstruction.java

```diff
@@ -145,22 +149,115 @@ else if(opcode.equalsIgnoreCase("transformapply"))
 			throw new DMLRuntimeException("Unknown opcode : " + opcode);
 		}
 	}
+	private void triangle(ExecutionContext ec, String opcode) {
+		boolean lower = opcode.equals("lowertri");
+		boolean diag = Boolean.parseBoolean(params.get("diag"));
+		boolean values = Boolean.parseBoolean(params.get("values"));
+
+		MatrixObject mo = (MatrixObject) getTarget(ec);
+
+		FederationMap fedMap = mo.getFedMapping();
+		boolean rowFed = mo.isFederated(FederationMap.FType.ROW);
+
+		long varID = FederationUtils.getNextFedDataID();
+		FederationMap diagFedMap;
+
+		diagFedMap = fedMap.mapParallel(varID, (range, data) -> {
+			try {
+				FederatedResponse response = data.executeFederatedOperation(new FederatedRequest(
+					FederatedRequest.RequestType.EXEC_UDF, -1,
+					new ParameterizedBuiltinFEDInstruction.Tri(data.getVarID(), varID,
+						rowFed ? (new int[] {range.getBeginDimsInt()[0], range.getEndDimsInt()[0]}) :
+							new int[] {range.getBeginDimsInt()[1], range.getEndDimsInt()[1]},
+						rowFed, lower, diag, values))).get();
+				if(!response.isSuccessful())
+					response.throwExceptionFromResponse();
+				return null;
+			}
+			catch(Exception e) {
+				throw new DMLRuntimeException(e);
+			}
+		});
+		MatrixObject out = ec.getMatrixObject(output);
+		out.setFedMapping(diagFedMap);
+	}
+
+	private static class Tri extends FederatedUDF {
+		private static final long serialVersionUID = 6254009025304038215L;
+
+		private final long _outputID;
+		private final int[] _slice;
+		private final boolean _rowFed;
+		private final boolean _lower;
+		private final boolean _diag;
+		private final boolean _values;
+
+		private Tri(long input, long outputID, int[] slice, boolean rowFed, boolean lower, boolean diag, boolean values) {
+			super(new long[] {input});
+			_outputID = outputID;
+			_slice = slice;
+			_rowFed = rowFed;
+			_lower = lower;
+			_diag = diag;
+			_values = values;
+		}
+
+		@Override
+		public FederatedResponse execute(ExecutionContext ec, Data... data) {
+			MatrixBlock mb = ((MatrixObject) data[0]).acquireReadAndRelease();
+			MatrixBlock soresBlock, addBlock;
+			MatrixBlock ret;
+
+			//slice
+			soresBlock = _rowFed ?
+				mb.slice(0, mb.getNumRows()-1, _slice[0], _slice[1]-1, new MatrixBlock()) :
+				mb.slice(_slice[0], _slice[1]-1);
+
+			//triangle
+			MatrixBlock tri = soresBlock.extractTriangular(new MatrixBlock(), _lower, _diag, _values);
+			if(_rowFed) {
+				ret = new MatrixBlock(mb.getNumRows(), mb.getNumColumns(), 0.0);
+				ret.copy(0, ret.getNumRows()-1, _slice[0], _slice[1]-1, tri, false);
+				if(_slice[1] <= mb.getNumColumns()-1 && !_lower) {
+					addBlock = mb.slice(0, mb.getNumRows()-1, _slice[1], mb.getNumColumns()-1, new MatrixBlock());
```

Comment:
the slice function is nice to use, but ultimately inefficient for the job you are doing here.
because the slice hash a large unnecessary overhead.
I would suggest extracting the sparse or dense block and processing that from the mb variable.

I would also suggest not to include this optimization right now in this PR, but add a todo for later.